activerecord/lib/active_record/load_tree.rb

```diff
@@ -0,0 +1,162 @@
+    def self.add_loaded_records(records)
+      return loaded_records unless ActiveRecord.dynamic_includes_overfetch_debugging?
+      return loaded_records if records.nil?
+      records = Array(records) unless records.is_a?(Array)
+      loaded_records.concat(records).uniq!
+    end
+
+    # All ActiveRecords instances marked as loaded.
+    def self.loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] ||= []
+    end
+
+    # Reset the state of tracked loaded and used records.
+    def self.clear_loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] = []
+    end
+
+    # Create a Load Tree instance representing the loaded record and its place in the overall
+    # hierarchy of the object tree. Any passed parents or sibilings should also include the load_tree
+    # instance on themselves.
+    #
+    # parent: Instance that is the parent of the loaded record.
+    # siblings: Instances that were loaded at the same time as the loaded record.
+    # association_name: The name of the association that was loaded.
+    def initialize(creator: nil, parent: nil, siblings: [], association_name: nil)
+      raise ArgumentError, "siblings must be an array" unless siblings.is_a?(Array)
+      raise ArgumentError, "Cannot be in an association if it is a root element" if parent.nil? && !association_name.nil?
+      raise ArgumentError, "all siblings must be have a load tree" unless siblings.all? { |r| r.respond_to?(:_load_tree) && r._load_tree.is_a?(LoadTree) }
+      set_parent(parent)
+
+      @record_class_name = creator.class.name
+      @association_name = association_name
+      parent._load_tree.add_loaded_association(association_name) unless parent.nil? || association_name.nil?
+      @loaded_associations = []
+      @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
+      reset_siblings(siblings)
+    end
+
+    def reset_siblings(siblings)
+      # We only want to keep sibling of the same class type as the creator
+      # This ensures we will not try to preload any associations that may not
+      # exist on some of the sibling which were loaded because of STI.
+      @siblings = siblings.select { |s| s.class.name == @record_class_name }.compact
+    end
+
+    # Does this load tree belong to a parent?
+    def root?
+      parent.nil?
+    end
+
+    # The full call path to the parent of this load tree. Traverses up the load tree
+    # checking if the current load tree is the the root, if it is then it returns
+    # the class name of the record for the root element, if not it returns the association
+    # name that was used to instantiate the current load tree's record. This will put together
+    # a full method chain.
+    #
+    # parent = Person.find(1)
+    # child = parent.children.first
+    # child._load_tree.full_load_call_path # => "Person.children"
+    #
+    def full_load_path
```

Comment:
`full_load_path` makes me think of the file load path, maybe `full_load_tree`? I'm not sure but I keep getting hung up on this name.

Reply:
I had not idea what to call this either, naming is hard... Part of me thinks it should be something having to do with a tree or a graph, because it is sort of like its place in a family tree, but I have no idea what that would be. Otherwise maybe something about methods called?